File: mobile/lib/features/auth/data/auth_models.dart
/// Wire models for the identity/auth endpoints, mirroring the backend DTOs in
/// `com.taarifu.identity.api.dto` exactly (AuthController).
///
/// Kept hand-written (no codegen) to keep the foundation lean; each parses the
/// `data` node the [ApiClient] hands it.
library;

/// The token pair returned by signup/login/refresh (backend `TokenPairDto`).
class TokenPair {
  /// Creates a pair.
  const TokenPair({required this.accessToken, required this.refreshToken});

  /// Short-lived JWT access token.
  final String accessToken;

  /// Rotating, single-use refresh token.
  final String refreshToken;

  /// Parses a `{ accessToken, refreshToken }` node.
  factory TokenPair.fromJson(Map<String, dynamic> json) => TokenPair(
    accessToken: json['accessToken'] as String,
    refreshToken: json['refreshToken'] as String,
  );
}

/// An OTP challenge id, returned by `/auth/otp/request` and the login-OTP
/// request (backend `OtpChallengeDto`). The client echoes it back on verify.
class OtpChallenge {
  /// Creates a challenge holder.
  const OtpChallenge({required this.challengeId});

  /// The opaque challenge id (UUID string) to present on verify.
  final String challengeId;

  /// Parses a `{ challengeId }` node.
  factory OtpChallenge.fromJson(Map<String, dynamic> json) =>
      OtpChallenge(challengeId: json['challengeId'] as String);
}

/// The result of completing signup (backend `AuthResultDto`): the new account's
/// public id, its trust [tier] (e.g. `T1`), and the issued [tokens].
class AuthResult {
  /// Creates a signup result.
  const AuthResult({
    required this.userPublicId,
    required this.tier,
    required this.tokens,
  });

  /// The account's public id (UUID string).
  final String userPublicId;

  /// The trust tier reached (e.g. `T1` after phone verification).
  final String tier;

  /// The freshly issued token pair.
  final TokenPair tokens;

  /// Parses a `{ userPublicId, tier, tokens }` node.
  factory AuthResult.fromJson(Map<String, dynamic> json) => AuthResult(
    userPublicId: json['userPublicId'] as String,
    tier: json['tier'] as String,
    tokens: TokenPair.fromJson(json['tokens'] as Map<String, dynamic>),
  );
}

/// The first-factor login outcome (backend `LoginResultDto`).
///
/// For a normal citizen [tokens] is populated and [mfaRequired] is false. For a
/// staff/MFA account [mfaRequired] is true, [tokens] is `null`, and [mfaToken]
/// must be completed at `/auth/login/totp` — the citizen app does not yet drive
/// that staff path, but the shape is modelled so it is not silently dropped.
class LoginResult {
  /// Creates a login outcome.
  const LoginResult({required this.mfaRequired, this.tokens, this.mfaToken});

  /// Whether a TOTP second factor is required (staff accounts).
  final bool mfaRequired;

  /// The issued pair, or `null` when [mfaRequired].
  final TokenPair? tokens;

  /// The MFA challenge token, or `null` when no MFA is needed.
  final String? mfaToken;

  /// Parses a `{ mfaRequired, tokens?, mfaToken? }` node.
  factory LoginResult.fromJson(Map<String, dynamic> json) => LoginResult(
    mfaRequired: json['mfaRequired'] == true,
    tokens: json['tokens'] == null
        ? null
        : TokenPair.fromJson(json['tokens'] as Map<String, dynamic>),
    mfaToken: json['mfaToken'] as String?,
  );
}

File: web-citizen/src/app/core/auth/auth.models.ts
/**
 * Auth domain types for the citizen PWA — mirror the backend auth DTOs (AuthController, AUTH-DESIGN §3/§5).
 *
 * <p>The citizen flow is **phone + OTP** (passwordless): request an OTP, then verify it to either complete
 * signup (new account → T1) or log in (existing account). Per the locked decision (one account / additive
 * roles) there is no separate "register vs login" identity — the same phone resolves to the same account.
 * The tier/role decoded from the JWT are a UI HINT ONLY — the server re-checks every protected action.</p>
 */

/** Body for `POST /auth/otp/request` (signup) and `POST /auth/login/otp/request` (login). */
export interface OtpRequest {
  /** Destination phone in E.164 (e.g. `+255700000001`). Validated client-side; never logged. */
  phone: string;
}

/** Response from an OTP request — the challenge id to verify against (carries no PII, anti-enumeration). */
export interface OtpChallenge {
  /** The OTP challenge public id, supplied back on verify. */
  challengeId: string;
}

/** Body for `POST /auth/signup` and `POST /auth/login/otp` — verifies a challenge with the SMS code. */
export interface VerifyOtpRequest {
  /** The challenge id returned by the request step. */
  challengeId: string;
  /** The numeric OTP the citizen received by SMS (never logged). */
  code: string;
}

/** The access + refresh JWT pair returned by signup/login/refresh (`TokenPairDto`). */
export interface TokenPair {
  /** Short-lived (~15 min) access JWT attached as `Authorization: Bearer`. */
  accessToken: string;
  /** Rotating, single-use refresh JWT (~30 days) exchanged at `/auth/refresh`. */
  refreshToken: string;
}

/** Result of a successful signup (`AuthResultDto`): account id, server tier hint, and the token pair. */
export interface AuthResult {
  /** The authenticated account's public id (UUID). */
  userPublicId: string;
  /** Server-computed trust tier name (UI hint; gating re-resolves live). */
  tier: string;
  /** The issued access + refresh pair. */
  tokens: TokenPair;
}

/**
 * Result of a first-factor login (`LoginResultDto`). For a citizen (no staff MFA) `tokens` is present and
 * `mfaRequired` is false. Staff/MFA accounts return `mfaRequired=true` with an `mfaToken` — the citizen
 * PWA does not drive the TOTP step; it surfaces a clear "use the staff console" message instead.
 */
export interface LoginResult {
  /** Whether a staff TOTP second factor is required (not applicable to citizen accounts). */
  mfaRequired: boolean;
  /** The issued token pair, or `null` when {@link mfaRequired} is true. */
  tokens: TokenPair | null;
  /** The MFA challenge token, or `null` when {@link tokens} are issued. */
  mfaToken: string | null;
}

/** Body for `POST /auth/refresh`. */
export interface RefreshRequest {
  /** The current refresh token to rotate. */
  refreshToken: string;
}

/**
 * The decoded, UI-facing session — derived from the access JWT's claims plus the stored tokens.
 *
 * <p>UI gating (showing a tier-unlock prompt, enabling "file a report") reads {@link roles} and
 * {@link tier} from here. This is convenience only: the backend's method-level `@PreAuthorize` +
 * `@RequiresTier` is the real gate. Tampering with these claims client-side cannot grant a real
 * privilege.</p>
 */
export interface Session {
  /** The account's public id (UUID) from the `sub` claim. */
  userId: string;
  /** Display name/subject, if the token carries one. */
  name?: string;
  /** Granted role names (e.g. `CITIZEN`, `REPRESENTATIVE`) — UI gating hint only. */
  roles: string[];
  /** Trust tier (e.g. `T1`–`T3`) — UI gating hint only. */
  tier?: string;
  /** Access-token expiry as epoch seconds (`exp` claim). */
  exp?: number;
}
